### PR TITLE
Filter UMR Calls

### PR DESCRIPTION
**Pull request by lappazos**

### CODE REVIEW

File: src/team_lib/mhba/xccl_mhba_collective.c [hunk 1]
```diff
@@ -24,8 +24,7 @@ xccl_status_t xccl_mhba_collective_init_base(xccl_coll_op_args_t   *coll_args,
     return XCCL_OK;
 }
 
-static xccl_status_t xccl_mhba_node_fanin(xccl_mhba_team_t     *team,
```
Comment (vspetrov) [inline src/team_lib/mhba/xccl_mhba_collective.c]: unnecessary change - just formatting. previous format was automatic "clang" based code style (preventing too long lines)

File: src/team_lib/mhba/xccl_mhba_collective.c [hunk 2]
```diff
@@ -39,7 +38,7 @@ static xccl_status_t xccl_mhba_node_fanin(xccl_mhba_team_t     *team,
                                    request); // no option for failure status
 
     if (team->node.sbgp->group_rank != team->node.asr_rank) {
-        *team->node.ops[index].my_ctrl = request->seq_num;
+        *team->node.ops[index].my_ctrl = request->self_change_flag ? -request->seq_num : request->seq_num;
```
Comment (vspetrov) [inline src/team_lib/mhba/xccl_mhba_collective.c]: what about "seq_num = 0 " ? do we start from 1 ? what happens if "wrap around is achieved" (seq_num becomes 0) ?

File: src/team_lib/mhba/xccl_mhba_collective.c [hunk 3]
```diff
@@ -48,10 +47,16 @@ static xccl_status_t xccl_mhba_node_fanin(xccl_mhba_team_t     *team,
...
+                if (*ctrl_v == -request->seq_num){
+                    request->buff_change_flag = 1;
+                }
+                else {
+                    return XCCL_INPROGRESS;
+                }
             }
         }
     }
+    request->buff_change_flag = request->self_change_flag ? 1 : request->buff_change_flag;
```
Comment (vspetrov) [inline src/team_lib/mhba/xccl_mhba_collective.c]: This line (59) in only relevant for ASR, right? If so, i would move it into asr conditional branch, otherwise when reading the code i have to think why non_asr rank sets this (and it may not be initialized for non-asr)

File: src/team_lib/mhba/xccl_mhba_collective.c [hunk 4]
```diff
@@ -48,10 +47,16 @@ static xccl_status_t xccl_mhba_node_fanin(xccl_mhba_team_t     *team,
             ctrl_v = (int *)((ptrdiff_t)team->node.ops[index].ctrl +
                              MHBA_CTRL_SIZE * i);
             if (*ctrl_v != request->seq_num) {
```
Comment (vspetrov) [inline src/team_lib/mhba/xccl_mhba_collective.c]: i would re-write this multi-condition like this:
```
if (*ctrl_v == -request->seq_num) {
    request->buff_change_flag = 1;
} else if (*ctrl_v != request->seq_num) {
    return XCCL_INPROGRESS;
}
```
It has less code and more readable imho.

File: src/team_lib/mhba/xccl_mhba_collective.c (hunk 3 above)
Comment (vspetrov) [inline src/team_lib/mhba/xccl_mhba_collective.c]: minor comment: i would rename "buff_change_flag" to smth that reflects more the meaning: like "need_mkey_update". It is not immediately clear what is "buffer change": it may be the same buffer pointer but different msg size and then we need to rebuild mkeys.

File: src/team_lib/mhba/xccl_mhba_collective.c [hunk 5]
```diff
@@ -63,24 +68,22 @@ static xccl_status_t xccl_mhba_reg_fanin_start(xccl_coll_task_t *task)
...
-    int dr_mem_access_flags = IBV_ACCESS_LOCAL_WRITE | IBV_ACCESS_REMOTE_WRITE;
 
     xccl_mhba_debug("register memory buffers");
 
     ucs_rcache_region_t* send_ptr;
     ucs_rcache_region_t* recv_ptr;
     if(UCS_OK != ucs_rcache_get(team->rcache, (void *)request->args.buffer_info.src_buffer,
                                 request->args.buffer_info.len * team->size,
-                                PROT_READ,&sr_mem_access_flags, &send_ptr)) {
+                                PROT_READ,&request->self_change_flag, &send_ptr)) {
```
Comment (vspetrov) [inline src/team_lib/mhba/xccl_mhba_collective.c]: codestyle: space after ','

File: src/team_lib/mhba/xccl_mhba_collective.c [hunk 6]
```diff
@@ -63,24 +68,22 @@ static xccl_status_t xccl_mhba_reg_fanin_start(xccl_coll_task_t *task)
...
         xccl_mhba_error("Failed to register send_bf memory (errno=%d)", errno);
         return XCCL_ERR_NO_RESOURCE;
     }
     request->send_rcache_region_p = xccl_rcache_ucs_get_reg_data(send_ptr);
 
     if(UCS_OK != ucs_rcache_get(team->rcache, (void *)request->args.buffer_info.dst_buffer,
                                 request->args.buffer_info.len * team->size,
-                                PROT_WRITE,&dr_mem_access_flags,&recv_ptr)) {
+                                PROT_WRITE,&request->self_change_flag,&recv_ptr)) {
```
Comment (vspetrov) [inline src/team_lib/mhba/xccl_mhba_collective.c]: codestyle: space after ','

File: src/team_lib/mhba/xccl_mhba_collective.c [hunk 7]
```diff
@@ -183,8 +186,10 @@ static xccl_status_t xccl_mhba_asr_barrier_start(xccl_coll_task_t *task)
     xccl_mhba_team_t     *team    = request->team;
     xccl_mhba_debug("asr barrier start");
 
-    // despite while statement, non blocking because have independent cq, will be finished in a finite time
-    xccl_mhba_populate_send_recv_mkeys(team, request);
+    if(request->buff_change_flag) {
+        // despite while statement, non blocking because have independent cq, will be finished in a finite time
```
Comment (vspetrov) [inline src/team_lib/mhba/xccl_mhba_collective.c]: comment is not longer meaningful - there is no "while" statement there

File: src/team_lib/mhba/xccl_mhba_team.c [hunk 8]
```diff
@@ -95,8 +95,9 @@ static ucs_status_t rcache_reg_mr(void *context, ucs_rcache_t *rcache,void *arg,
     size_t length             = (size_t)(rregion->super.end - rregion->super.start);
     xccl_mhba_reg_t* mhba_reg = xccl_rcache_ucs_get_reg_data(rregion);
     mhba_reg->region = rregion;
-    int* mem_flags = (int*) arg;
-    mhba_reg->mr = ibv_reg_mr(team->node.shared_pd, addr, length, *mem_flags);
+    int* change_flag = (int*) arg;
+    *change_flag = 1;
```
Comment (vspetrov) [inline src/team_lib/mhba/xccl_mhba_team.c:99]: So, you decide if you need to update the key only based on the rcache registration callback execution? I don't get it: how it works? If the buffer was registered and not released by user - it is still in rcache. But the mkey is not longer relevant. Here is an example:
1. user calls alltoall with buffer = BUF1 (assume msgsize = 4 for simplicity)
2. BUF1 will get into UCS rcache. Mkey is created for BUF1/4b
3. Now, user calls alltoall with BUF2 (assume same 4b)
4. BUF2 gets into UCS rcache as well. Mkey is now created for BUF2/4b
5. Not user may call alltoall with BUF1 again, it will be found in UCS rcache and hence registration callback is not called, you don't set the change_flag. Thus, you will use wrong Mkey from BUF2/4b.

Second question: how will you handle the case when BUF1 is called second time but with different msgsize (say, 2bytes). It will be found in rcache but Mkey is wrong.

Need a better logic. You need to write some kind of function that builds a "key" based on buffer/msgsize and then check if the "key" from previous alltoall call is exactly same as next one. Only in this case you can re-use mkey, otherwise need an update.

**Review by vspetrov — CHANGES_REQUESTED** [review]

File: src/team_lib/mhba/xccl_mhba_collective.c (hunk 2 above)
Comment (lappazos) [reply src/team_lib/mhba/xccl_mhba_collective.c:41]: in 0 we'll always have to update, so line 59 takes care of it. because for asr self_change flag will always be 1 for 0. ill add a comment explainig this

Comment (lappazos) [reply src/team_lib/mhba/xccl_mhba_collective.c]: after the fix, seq_num 0 won't matter because msg_size will be different, as it set to 0 in the beginning

File: src/team_lib/mhba/xccl_mhba_team.c (hunk 8 above)
Comment (lappazos) [reply src/team_lib/mhba/xccl_mhba_team.c:99]: changed

**Review by vspetrov — CHANGES_REQUESTED** [review]
I still think PR is incorrect. Now you added the ptr check but removed ibv_reg_mr change flag (from rcache_get). Suppose you have a buffer that users passes with the address PTR1. We register it and put into a cache. User then does free and malloc again - malloc can likely return the same PTR1 but it will point to another physical page (and the registration in the rcache will be revoked). So, you ptr check will be wrong.

We need to have 3 values key + reg_mr check flag:
typedef struct xccl_mhba_buffer_key {
    void *sbuf;
    void *rbuf;
    size_t len;
    int block_size; // layout of UMR also depends on the BS
} xccl_mhba_buffer_key_t;

You will store "previous" buffer_key on the team.
When collective is starting you init 2 fields:
int registration_changed;
ucs_racache_get ( ... , &registration_changed) - similar to what you had in the last version
xccl_mhba_buffer_key_t = {
   .sbuf = sbuf,
   .rbuf  =rbuf,
   .len = len,
    .block_size = block_size}
This key can be initiazlized during "collective_init".
THen you can write a simple "inline" function
int xccl_mhba_need_mkey_update() {
    if (registration_changed || COMPARE_BUFFER_KEY(buffer_key,previous_buffer_key)) {
return 1;
} else {
return 0;
}
}

where COMPARE_KEY is a macro that compares structs.

Smth like that. Makes sense ?
